File: osakamenesu/apps/web/src/lib/dashboard-therapists.ts
import { buildApiUrl, resolveApiBases } from '@/lib/api'
import type { DashboardShopRequestOptions } from './dashboard-shops'

export type DashboardTherapistStatus = 'draft' | 'published' | 'archived'

export type DashboardTherapistSummary = {
  id: string
  name: string
  alias?: string | null
  headline?: string | null
  status: DashboardTherapistStatus
  display_order: number
  is_booking_enabled: boolean
  updated_at: string
  photo_urls: string[]
  specialties: string[]
}

export type DashboardTherapistDetail = DashboardTherapistSummary & {
  biography?: string | null
  qualifications: string[]
  experience_years?: number | null
  created_at: string
}

export type DashboardTherapistCreatePayload = {
  name: string
  alias?: string | null
  headline?: string | null
  biography?: string | null
  specialties?: string[]
  qualifications?: string[]
  experience_years?: number | null
  photo_urls?: string[]
  is_booking_enabled?: boolean
}

export type DashboardTherapistUpdatePayload = {
  updated_at: string
  name?: string
  alias?: string | null
  headline?: string | null
  biography?: string | null
  specialties?: string[]
  qualifications?: string[]
  experience_years?: number | null
  photo_urls?: string[]
  status?: DashboardTherapistStatus
  is_booking_enabled?: boolean
  display_order?: number
}

export type DashboardTherapistReorderPayload = {
  items: {
    therapist_id: string
    display_order: number
  }[]
}

export type DashboardTherapistListResult =
  | { status: 'success'; data: DashboardTherapistSummary[] }
  | { status: 'unauthorized' }
  | { status: 'forbidden'; detail?: string }
  | { status: 'not_found' }
  | { status: 'error'; message: string }

export type DashboardTherapistMutationResult =
  | { status: 'success'; data: DashboardTherapistDetail }
  | { status: 'validation_error'; detail: unknown }
  | { status: 'conflict'; current: DashboardTherapistDetail }
  | { status: 'unauthorized' }
  | { status: 'forbidden'; detail?: string }
  | { status: 'not_found' }
  | { status: 'error'; message: string }

export type DashboardTherapistDeleteResult =
  | { status: 'success' }
  | { status: 'unauthorized' }
  | { status: 'forbidden'; detail?: string }
  | { status: 'not_found' }
  | { status: 'error'; message: string }

function createRequestInit(
  method: string,
  options?: DashboardShopRequestOptions,
  body?: unknown
): RequestInit {
  const headers: Record<string, string> = {}
  if (options?.cookieHeader) {
    headers.cookie = options.cookieHeader
  }
  if (body !== undefined) {
    headers['Content-Type'] = 'application/json'
  }

  const init: RequestInit = {
    method,
    cache: options?.cache ?? 'no-store',
    signal: options?.signal,
    credentials: options?.cookieHeader ? 'omit' : 'include',
  }

  if (Object.keys(headers).length) {
    init.headers = headers
  }

  if (body !== undefined) {
    init.body = JSON.stringify(body)
  }

  return init
}

async function requestJson<T>(
  path: string,
  init: RequestInit,
  successStatuses: number[]
): Promise<{ response: Response; data?: T }> {
  let lastError: { status: string; message?: string } | null = null

  for (const base of resolveApiBases()) {
    try {
      const res = await fetch(buildApiUrl(base, path), init)

      if (successStatuses.includes(res.status)) {
        let data: T | undefined
        if (res.status !== 204 && res.headers.get('content-type')?.includes('json')) {
          data = (await res.json()) as T
        }
        return { response: res, data }
      }

      if ([401, 403, 404, 409, 422].includes(res.status)) {
        let data: T | undefined
        if (res.headers.get('content-type')?.includes('json')) {
          data = (await res.json()) as T
        }
        return { response: res, data }
      }

      lastError = {
        status: 'error',
        message: `リクエストに失敗しました (status=${res.status})`,
      }
    } catch (error) {
      lastError = {
        status: 'error',
        message: error instanceof Error ? error.message : 'リクエスト中にエラーが発生しました',
      }
    }
  }

  throw lastError ?? {
    status: 'error',
    message: 'API リクエストが完了しませんでした',
  }
}

function toSummary(detail: DashboardTherapistDetail): DashboardTherapistSummary {
  return {
    id: detail.id,
    name: detail.name,
    alias: detail.alias ?? null,
    headline: detail.headline ?? null,
    status: detail.status,
    display_order: detail.display_order,
    is_booking_enabled: detail.is_booking_enabled,
    updated_at: detail.updated_at,
    photo_urls: Array.isArray(detail.photo_urls) ? detail.photo_urls : [],
    specialties: Array.isArray(detail.specialties) ? detail.specialties : [],
  }
}

export function summarizeTherapist(detail: DashboardTherapistDetail): DashboardTherapistSummary {
  return toSummary(detail)
}

export async function fetchDashboardTherapists(
  profileId: string,
  options?: DashboardShopRequestOptions
): Promise<DashboardTherapistListResult> {
  try {
    const { response, data } = await requestJson<DashboardTherapistSummary[]>(
      `api/dashboard/shops/${profileId}/therapists`,
      createRequestInit('GET', options),
      [200]
    )

    switch (response.status) {
      case 200:
        return { status: 'success', data: data ?? [] }
      case 401:
        return { status: 'unauthorized' }
      case 403:
        return {
          status: 'forbidden',
          detail: (data as { detail?: string } | undefined)?.detail,
        }
      case 404:
        return { status: 'not_found' }
      default:
        return {
          status: 'error',
          message: `セラピスト情報の取得に失敗しました (status=${response.status})`,
        }
    }
  } catch (error) {
    if (typeof error === 'object' && error !== null && 'status' in error) {
      return error as DashboardTherapistListResult
    }
    return {
      status: 'error',
      message: error instanceof Error ? error.message : 'セラピスト情報の取得に失敗しました',
    }
  }
}

export async function fetchDashboardTherapist(
  profileId: string,
  therapistId: string,
  options?: DashboardShopRequestOptions
): Promise<DashboardTherapistMutationResult> {
  try {
    const { response, data } = await requestJson<DashboardTherapistDetail>(
      `api/dashboard/shops/${profileId}/therapists/${therapistId}`,
      createRequestInit('GET', options),
      [200]
    )

    switch (response.status) {
      case 200:
        return { status: 'success', data: data! }
      case 401:
        return { status: 'unauthorized' }
      case 403:
        return {
          status: 'forbidden',
          detail: (data as { detail?: string } | undefined)?.detail,
        }
      case 404:
        return { status: 'not_found' }
      default:
        return {
          status: 'error',
          message: `セラピスト情報の取得に失敗しました (status=${response.status})`,
        }
    }
  } catch (error) {
    if (typeof error === 'object' && error !== null && 'status' in error) {
      return error as DashboardTherapistMutationResult
    }
    return {
      status: 'error',
      message: error instanceof Error ? error.message : 'セラピスト情報の取得に失敗しました',
    }
  }
}

export async function createDashboardTherapist(
  profileId: string,
  payload: DashboardTherapistCreatePayload,
  options?: DashboardShopRequestOptions
): Promise<DashboardTherapistMutationResult> {
  try {
    const { response, data } = await requestJson<DashboardTherapistDetail>(
      `api/dashboard/shops/${profileId}/therapists`,
      createRequestInit('POST', options, payload),
      [201]
    )

    switch (response.status) {
      case 201:
        return { status: 'success', data: data! }
      case 401:
        return { status: 'unauthorized' }
      case 403:
        return {
          status: 'forbidden',
          detail: (data as { detail?: string } | undefined)?.detail,
        }
      case 422:
        return { status: 'validation_error', detail: data }
      default:
        return {
          status: 'error',
          message: `セラピストの作成に失敗しました (status=${response.status})`,
        }
    }
  } catch (error) {
    if (typeof error === 'object' && error !== null && 'status' in error) {
      return error as DashboardTherapistMutationResult
    }
    return {
      status: 'error',
      message: error instanceof Error ? error.message : 'セラピストの作成に失敗しました',
    }
  }
}

export async function updateDashboardTherapist(
  profileId: string,
  therapistId: string,
  payload: DashboardTherapistUpdatePayload,
  options?: DashboardShopRequestOptions
): Promise<DashboardTherapistMutationResult> {
  try {
    const { response, data } = await requestJson<DashboardTherapistDetail>(
      `api/dashboard/shops/${profileId}/therapists/${therapistId}`,
      createRequestInit('PATCH', options, payload),
      [200]
    )

    switch (response.status) {
      case 200:
        return { status: 'success', data: data! }
      case 401:
        return { status: 'unauthorized' }
      case 403:
        return {
          status: 'forbidden',
          detail: (data as { detail?: string } | undefined)?.detail,
        }
      case 404:
        return { status: 'not_found' }
      case 409:
        return { status: 'conflict', current: data! }
      case 422:
        return { status: 'validation_error', detail: data }
      default:
        return {
          status: 'error',
          message: `セラピストの更新に失敗しました (status=${response.status})`,
        }
    }
  } catch (error) {
    if (typeof error === 'object' && error !== null && 'status' in error) {
      return error as DashboardTherapistMutationResult
    }
    return {
      status: 'error',
      message: error instanceof Error ? error.message : 'セラピストの更新に失敗しました',
    }
  }
}

export async function deleteDashboardTherapist(
  profileId: string,
  therapistId: string,
  options?: DashboardShopRequestOptions
): Promise<DashboardTherapistDeleteResult> {
  try {
    const { response } = await requestJson<undefined>(
      `api/dashboard/shops/${profileId}/therapists/${therapistId}`,
      createRequestInit('DELETE', options),
      [204]
    )

    switch (response.status) {
      case 204:
        return { status: 'success' }
      case 401:
        return { status: 'unauthorized' }
      case 403:
        return { status: 'forbidden' }
      case 404:
        return { status: 'not_found' }
      default:
        return {
          status: 'error',
          message: `セラピストの削除に失敗しました (status=${response.status})`,
        }
    }
  } catch (error) {
    if (typeof error === 'object' && error !== null && 'status' in error) {
      return error as DashboardTherapistDeleteResult
    }
    return {
      status: 'error',
      message: error instanceof Error ? error.message : 'セラピストの削除に失敗しました',
    }
  }
}

export async function reorderDashboardTherapists(
  profileId: string,
  payload: DashboardTherapistReorderPayload,
  options?: DashboardShopRequestOptions
): Promise<DashboardTherapistListResult> {
  try {
    const { response, data } = await requestJson<DashboardTherapistSummary[]>(
      `api/dashboard/shops/${profileId}/therapists:reorder`,
      createRequestInit('POST', options, payload),
      [200]
    )

    switch (response.status) {
      case 200:
        return { status: 'success', data: data ?? [] }
      case 401:
        return { status: 'unauthorized' }
      case 403:
        return {
          status: 'forbidden',
          detail: (data as { detail?: string } | undefined)?.detail,
        }
      case 404:
        return { status: 'not_found' }
      default:
        return {
          status: 'error',
          message: `セラピストの並び替えに失敗しました (status=${response.status})`,
        }
    }
  } catch (error) {
    if (typeof error === 'object' && error !== null && 'status' in error) {
      return error as DashboardTherapistListResult
    }
    return {
      status: 'error',
      message: error instanceof Error ? error.message : 'セラピストの並び替えに失敗しました',
    }
  }
}

File: osakamenesu/services/api/app/routers/dashboard_therapists.py
from __future__ import annotations

import hashlib
from datetime import datetime, timezone
from typing import Any, Iterable
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Request, Response, status
import sqlalchemy as sa
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from .. import models
from ..db import get_session
from ..deps import require_user
from ..schemas import (
    DashboardTherapistCreatePayload,
    DashboardTherapistDetail,
    DashboardTherapistReorderPayload,
    DashboardTherapistSummary,
    DashboardTherapistUpdatePayload,
)
from ..utils.profiles import build_profile_doc
from zoneinfo import ZoneInfo

router = APIRouter(prefix="/api/dashboard", tags=["dashboard-therapists"])

JST = ZoneInfo("Asia/Tokyo")


def _ensure_datetime(value: datetime) -> datetime:
    if value.tzinfo is None:
        return value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc)


async def _get_profile(db: AsyncSession, profile_id: UUID) -> models.Profile:
    profile = await db.get(models.Profile, profile_id)
    if not profile:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="profile_not_found")
    return profile


def _sanitize_strings(values: Iterable[str] | None) -> list[str]:
    if not values:
        return []
    sanitized: list[str] = []
    for value in values:
        if not isinstance(value, str):
            continue
        text = value.strip()
        if text:
            sanitized.append(text)
    return sanitized


def _sanitize_photo_urls(values: Iterable[str] | None) -> list[str]:
    return _sanitize_strings(values)


def _serialize_therapist(model: models.Therapist) -> DashboardTherapistDetail:
    specialties = list(model.specialties or [])
    qualifications = list(model.qualifications or [])
    photo_urls = list(model.photo_urls or [])
    return DashboardTherapistDetail(
        id=model.id,
        name=model.name,
        alias=model.alias,
        headline=model.headline,
        biography=model.biography,
        specialties=[str(item) for item in specialties],
        qualifications=[str(item) for item in qualifications],
        experience_years=model.experience_years,
        status=model.status,
        display_order=model.display_order,
        is_booking_enabled=model.is_booking_enabled,
        photo_urls=[str(url) for url in photo_urls],
        created_at=model.created_at,
        updated_at=model.updated_at,
    )


def _summary_from_detail(detail: DashboardTherapistDetail) -> DashboardTherapistSummary:
    return DashboardTherapistSummary(
        id=detail.id,
        name=detail.name,
        alias=detail.alias,
        headline=detail.headline,
        status=detail.status,
        display_order=detail.display_order,
        is_booking_enabled=detail.is_booking_enabled,
        updated_at=detail.updated_at,
        photo_urls=detail.photo_urls,
        specialties=detail.specialties,
    )


async def _reindex_profile(db: AsyncSession, profile: models.Profile) -> None:
    today = datetime.now(JST).date()
    availability_count = await db.execute(
        select(sa.func.count())
        .select_from(models.Availability)
        .where(models.Availability.profile_id == profile.id, models.Availability.date == today)
    )
    has_today = (availability_count.scalar_one() or 0) > 0
    outlinks = await db.execute(select(models.Outlink).where(models.Outlink.profile_id == profile.id))
    doc = build_profile_doc(
        profile,
        today=has_today,
        tag_score=0.0,
        ctr7d=0.0,
        outlinks=list(outlinks.scalars().all()),
    )
    try:
        from ..meili import index_profile

        index_profile(doc)
    except Exception:
        # Non-blocking: メイリーダウンで編集を失敗にしない
        pass


async def _sync_staff_contact_json(db: AsyncSession, profile: models.Profile) -> None:
    result = await db.execute(
        select(models.Therapist)
        .where(models.Therapist.profile_id == profile.id)
        .order_by(models.Therapist.display_order, models.Therapist.created_at)
    )
    therapists = list(result.scalars().all())
    staff_payload: list[dict[str, Any]] = []
    for therapist in therapists:
        staff_payload.append(
            {
                "id": str(therapist.id),
                "name": therapist.name,
                "alias": therapist.alias,
                "headline": therapist.headline,
                "biography": therapist.biography,
                "specialties": list(therapist.specialties or []),
                "qualifications": list(therapist.qualifications or []),
                "experience_years": therapist.experience_years,
                "photo_urls": list(therapist.photo_urls or []),
                "display_order": therapist.display_order,
                "status": therapist.status,
                "is_booking_enabled": therapist.is_booking_enabled,
            }
        )
    contact_json = dict(profile.contact_json or {})
    if staff_payload:
        contact_json["staff"] = staff_payload
    else:
        contact_json.pop("staff", None)
    profile.contact_json = contact_json
    await db.flush([profile])


async def _record_change(
    request: Request,
    db: AsyncSession,
    target_id: UUID | None,
    action: str,
    before: Any,
    after: Any,
) -> None:
    try:
        ip = request.headers.get("x-forwarded-for") or (request.client.host if request.client else "")
        ip_hash = hashlib.sha256(ip.encode("utf-8")).hexdigest() if ip else None
        log = models.AdminChangeLog(
            target_type="therapist",
            target_id=target_id,
            action=action,
            before_json=before,
            after_json=after,
            admin_key_hash=None,
            ip_hash=ip_hash,
        )
        db.add(log)
        await db.commit()
    except Exception:
        # 監査ログは失敗しても処理を止めない
        pass


@router.get(
    "/shops/{profile_id}/therapists",
    response_model=list[DashboardTherapistSummary],
)
async def list_dashboard_therapists(
    profile_id: UUID,
    db: AsyncSession = Depends(get_session),
    user: models.User = Depends(require_user),
) -> list[DashboardTherapistSummary]:
    _ = user
    await _get_profile(db, profile_id)
    result = await db.execute(
        select(models.Therapist)
        .where(models.Therapist.profile_id == profile_id)
        .order_by(models.Therapist.display_order, models.Therapist.created_at)
    )
    therapists = list(result.scalars().all())
    summaries = []
    for therapist in therapists:
        detail = _serialize_therapist(therapist)
        summaries.append(_summary_from_detail(detail))
    return summaries


@router.post(
    "/shops/{profile_id}/therapists",
    response_model=DashboardTherapistDetail,
    status_code=status.HTTP_201_CREATED,
)
async def create_dashboard_therapist(
    request: Request,
    profile_id: UUID,
    payload: DashboardTherapistCreatePayload,
    db: AsyncSession = Depends(get_session),
    user: models.User = Depends(require_user),
) -> DashboardTherapistDetail:
    _ = user
    profile = await _get_profile(db, profile_id)
    name = payload.name.strip() if payload.name else ""
    if not name:
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail={"field": "name", "message": "セラピスト名を入力してください。"},
        )

    specialties = _sanitize_strings(payload.specialties)
    qualifications = _sanitize_strings(payload.qualifications)
    photo_urls = _sanitize_photo_urls(payload.photo_urls)
    experience_years = None
    if payload.experience_years is not None:
        experience_years = max(0, int(payload.experience_years))

    display_order_res = await db.execute(
        select(models.Therapist.display_order)
        .where(models.Therapist.profile_id == profile_id)
        .order_by(models.Therapist.display_order.desc())
    )
    next_order = 0
    top = display_order_res.first()
    if top:
        next_order = (top[0] or 0) + 10

    therapist = models.Therapist(
        profile_id=profile_id,
        name=name,
        alias=payload.alias.strip() if payload.alias else None,
        headline=payload.headline.strip() if payload.headline else None,
        biography=payload.biography.strip() if payload.biography else None,
        specialties=specialties or None,
        qualifications=qualifications or None,
        experience_years=experience_years,
        photo_urls=photo_urls or None,
        is_booking_enabled=payload.is_booking_enabled if payload.is_booking_enabled is not None else True,
        display_order=next_order,
        status="draft",
    )
    db.add(therapist)
    await db.flush()
    await _sync_staff_contact_json(db, profile)
    await db.commit()
    await db.refresh(therapist)

    await _reindex_profile(db, profile)

    detail = _serialize_therapist(therapist)
    await _record_change(request, db, detail.id, "create", None, detail.model_dump())
    return detail


@router.get(
    "/shops/{profile_id}/therapists/{therapist_id}",
    response_model=DashboardTherapistDetail,
)
async def get_dashboard_therapist(
    profile_id: UUID,
    therapist_id: UUID,
    db: AsyncSession = Depends(get_session),
    user: models.User = Depends(require_user),
) -> DashboardTherapistDetail:
    _ = user
    await _get_profile(db, profile_id)
    therapist = await db.get(models.Therapist, therapist_id)
    if not therapist or therapist.profile_id != profile_id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="therapist_not_found")
    return _serialize_therapist(therapist)


@router.patch(
    "/shops/{profile_id}/therapists/{therapist_id}",
    response_model=DashboardTherapistDetail,
)
async def update_dashboard_therapist(
    request: Request,
    profile_id: UUID,
    therapist_id: UUID,
    payload: DashboardTherapistUpdatePayload,
    db: AsyncSession = Depends(get_session),
    user: models.User = Depends(require_user),
) -> DashboardTherapistDetail:
    _ = user
    profile = await _get_profile(db, profile_id)
    therapist = await db.get(models.Therapist, therapist_id)
    if not therapist or therapist.profile_id != profile_id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="therapist_not_found")

    current_updated_at = _ensure_datetime(therapist.updated_at)
    incoming_updated_at = _ensure_datetime(payload.updated_at)
    if incoming_updated_at != current_updated_at:
        raise HTTPException(
            status.HTTP_409_CONFLICT,
            detail={
                "message": "conflict",
                "current": _serialize_therapist(therapist).model_dump(),
            },
        )

    before = _serialize_therapist(therapist).model_dump()

    if payload.name is not None:
        name = payload.name.strip()
        if not name:
            raise HTTPException(
                status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail={"field": "name", "message": "セラピスト名を入力してください。"},
            )
        therapist.name = name

    if payload.alias is not None:
        therapist.alias = payload.alias.strip() or None

    if payload.headline is not None:
        therapist.headline = payload.headline.strip() or None

    if payload.biography is not None:
        therapist.biography = payload.biography.strip() or None

    if payload.specialties is not None:
        therapist.specialties = _sanitize_strings(payload.specialties) or None

    if payload.qualifications is not None:
        therapist.qualifications = _sanitize_strings(payload.qualifications) or None

    if payload.experience_years is not None:
        therapist.experience_years = max(0, int(payload.experience_years))

    if payload.photo_urls is not None:
        therapist.photo_urls = _sanitize_photo_urls(payload.photo_urls) or None

    if payload.status is not None:
        therapist.status = payload.status

    if payload.is_booking_enabled is not None:
        therapist.is_booking_enabled = bool(payload.is_booking_enabled)

    if payload.display_order is not None:
        therapist.display_order = max(0, int(payload.display_order))

    await db.flush()
    await _sync_staff_contact_json(db, profile)
    await db.commit()
    await db.refresh(therapist)
    await _reindex_profile(db, profile)

    detail = _serialize_therapist(therapist)
    await _record_change(request, db, detail.id, "update", before, detail.model_dump())
    return detail


@router.delete(
    "/shops/{profile_id}/therapists/{therapist_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_class=Response,
)
async def delete_dashboard_therapist(
    request: Request,
    profile_id: UUID,
    therapist_id: UUID,
    db: AsyncSession = Depends(get_session),
    user: models.User = Depends(require_user),
) -> Response:
    _ = user
    profile = await _get_profile(db, profile_id)
    therapist = await db.get(models.Therapist, therapist_id)
    if not therapist or therapist.profile_id != profile_id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="therapist_not_found")

    before = _serialize_therapist(therapist).model_dump()
    await db.delete(therapist)
    await db.flush()

    await _sync_staff_contact_json(db, profile)
    await db.commit()
    await _reindex_profile(db, profile)
    await _record_change(request, db, therapist.id, "delete", before, None)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post(
    "/shops/{profile_id}/therapists:reorder",
    response_model=list[DashboardTherapistSummary],
)
async def reorder_dashboard_therapists(
    request: Request,
    profile_id: UUID,
    payload: DashboardTherapistReorderPayload,
    db: AsyncSession = Depends(get_session),
    user: models.User = Depends(require_user),
) -> list[DashboardTherapistSummary]:
    _ = user
    profile = await _get_profile(db, profile_id)
    if not payload.items:
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail={"message": "items_required"},
        )

    therapist_ids = {item.therapist_id for item in payload.items}
    result = await db.execute(
        select(models.Therapist)
        .where(models.Therapist.profile_id == profile_id, models.Therapist.id.in_(therapist_ids))
    )
    existing = {therapist.id: therapist for therapist in result.scalars().all()}
    if len(existing) != len(therapist_ids):
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="therapist_not_found")

    before_state = [
        _serialize_therapist(existing[tid]).model_dump()
        for tid in sorted(existing, key=lambda t: existing[t].display_order)
    ]

    for item in payload.items:
        therapist = existing[item.therapist_id]
        therapist.display_order = max(0, item.display_order)

    await db.flush()
    await _sync_staff_contact_json(db, profile)
    await db.commit()

    result = await db.execute(
        select(models.Therapist)
        .where(models.Therapist.profile_id == profile_id)
        .order_by(models.Therapist.display_order, models.Therapist.created_at)
    )
    therapists = list(result.scalars().all())
    summaries = [_summary_from_detail(_serialize_therapist(t)) for t in therapists]

    after_state = [summary.model_dump() for summary in summaries]
    await _reindex_profile(db, profile)
    await _record_change(request, db, None, "reorder", before_state, after_state)
    return summaries
